app/controllers/key_pairs_controller.rb

```diff
@@ -38,36 +49,91 @@ def edit
+  # is used to create the public key
   # Creates the KEY_STORAGE directory if it does not yet exist
-  def upload_key_file(_file, user_name)
+  def upload_key_file(isFile, _file, time_stamp)
     Dir.mkdir(KEY_STORAGE) unless File.exists?(KEY_STORAGE)
-    uploaded_io = _file
 
-    # todo: check current user to see if they have permission to
-    # use a different user_name then their current one (admins)
+    public_key_content = ''
+
+    if !isFile
+      public_key_content = _file
+    else
+      public_key_content = _file.read
+    end
 
-    File.open(Rails.root.join(KEY_STORAGE, user_name + '.pub'), 'wb') do |f|
-      f.write(uploaded_io.read)
+    File.open(Rails.root.join(KEY_STORAGE, @current_user.user_name + "@" +
+                                             time_stamp + '.pub'), 'wb') do |f|
+      f.write(public_key_content)
     end
+
+    add_key(KEY_STORAGE + '/' + @current_user.user_name + "@" + time_stamp +
+                '.pub')
+  end
+
+  # Adds a specific public key to a specific user.
+  def add_key(_path)
+    ga_repo = Gitolite::GitoliteAdmin.new(
+        REPOSITORY_STORAGE + '/gitolite-admin', GITOLITE_SETTINGS)
+
+    # Check to see if an individual repo exists for this user
+    key = Gitolite::SSHKey.from_file(_path)
+
+    ga_repo.add_key(key)
+
+    adminKey = Gitolite::SSHKey.from_file(GITOLITE_SETTINGS[:public_key])
+    ga_repo.add_key(adminKey)
+
+    # update Gitolite repo
+    ga_repo.save_and_apply
+  end
+
+  # Deletes a specific public key from a specific user.
+  def remove_key(_path)
+
+    ga_repo = Gitolite::GitoliteAdmin.new(
+        REPOSITORY_STORAGE + '/gitolite-admin', GITOLITE_SETTINGS)
+
+    # Check to see if an individual repo exists for this user
+    key = Gitolite::SSHKey.from_file(_path)
+
+    ga_repo.rm_key(key)
+
+    adminKey = Gitolite::SSHKey.from_file(GITOLITE_SETTINGS[:public_key])
+    ga_repo.add_key(adminKey)
+
+    # update Gitolite repo
+    ga_repo.save_and_apply
   end
 
   # POST /key_pairs
   # POST /key_pairs.json
   def create
-    # Upload the file
-    upload_key_file(key_pair_params[:file], @current_user.user_name)
+    # Used to uniquely identify key
+    time_stamp = Time.now.to_i.to_s
+
+    # If user uploads the public key as a file then that takes precedence over
+    # the key string
+    if !key_pair_params[:file]
+      # Create a .pub file on the file system
+      upload_key_file(false, key_pair_params[:key_string], time_stamp)
+    else
+      # Upload the file
+      upload_key_file(true, key_pair_params[:file], time_stamp)
+    end
 
     # Save the record
     @key_pair = KeyPair.new(
       key_pair_params.merge(user_name: @current_user.user_name,
-                            file_name: @current_user.user_name + '.pub'))
+                            user_id: @current_user.id,
+                            file_name: @current_user.user_name +
+                                "@" + time_stamp + '.pub'))
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.